lib/commands/streamscreen.js

```diff
@@ -0,0 +1,405 @@
+import _ from 'lodash';
+import { fs, system } from 'appium-support';
+import log from '../logger';
+import { exec, SubProcess } from 'teen_process';
+import { checkPortStatus } from 'portscanner';
+import http from 'http';
+import net from 'net';
+import B from 'bluebird';
+import { waitForCondition } from 'asyncbox';
+import { spawn } from 'child_process';
+import { quote } from 'shell-quote';
+
+const commands = {};
+
+const RECORDING_INTERVAL_SEC = 5;
+const STREAMING_STARTUP_TIMEOUT_MS = 5000;
+const GSTREAMER_BINARY = `gst-launch-1.0${system.isWindows() ? '.exe' : ''}`;
+const GST_INSPECT_BINARY = `gst-inspect-1.0${system.isWindows() ? '.exe' : ''}`;
+const REQUIRED_GST_PLUGINS = {
+  avdec_h264: 'gst-libav',
+  h264parse: 'gst-plugins-bad',
+  jpegenc: 'gst-plugins-good',
+  tcpserversink: 'gst-plugins-base',
+  multipartmux: 'gst-plugins-good',
+};
+const SCREENRECORD_BINARY = 'screenrecord';
+const GST_TUTORIAL_URL = 'https://gstreamer.freedesktop.org/documentation/installing/index.html';
+const DEFAULT_HOST = '127.0.0.1';
+const TCP_HOST = '127.0.0.1';
+const DEFAULT_PORT = 8093;
+const DEFAULT_QUALITY = 70;
+const DEFAULT_BITRATE = 4000000; // Mbps
+const BOUNDARY_STRING = '--2ae9746887f170b8cf7c271047ce314c';
+
+
+async function verifyStreamingRequirements (adb) {
+  if (!_.trim(await adb.shell(['which', SCREENRECORD_BINARY]))) {
+    throw new Error(
+      `The required '${SCREENRECORD_BINARY}' binary is not available on the device under test`);
+  }
+
+  for (const binaryName of [GSTREAMER_BINARY, GST_INSPECT_BINARY]) {
+    try {
+      await fs.which(binaryName);
+    } catch (e) {
+      throw new Error(`The ${binaryName} is not available in PATH on the host system. ` +
+        `See ${GST_TUTORIAL_URL} for more details on how to install it.`);
+    }
+  }
+
+  for (const [name, modName] of _.toPairs(REQUIRED_GST_PLUGINS)) {
+    const {stdout} = await exec(GST_INSPECT_BINARY, [name]);
+    if (!_.includes(stdout, modName)) {
+      throw new Error(
+        `The required GStreamer plugin '${name}' from '${modName}' module is not installed. ` +
+        `See ${GST_TUTORIAL_URL} for more details on how to install it.`);
+    }
+  }
+}
+
+async function getDeviceInfo (adb) {
+  const output = await adb.shell(['dumpsys', 'display']);
+  const widthMatch = /\bdeviceWidth=(\d+)/.exec(output);
+  if (!widthMatch) {
+    throw new Error(`Cannot parse the device width from ${output}`);
```

Comment:
If `output` is in single quotes it would be easier to see if there is oddities in whitespace in the output, when an error occurs.

Reply:
the output here is quite long multiline string. How would you advice to print it?

Reply:
As it stands its all being added to the error. I just think it is good to have some sort of delimiter in case something is off. There have been cases where whitespace has mattered and it's been unclear in the logging.

Reply:
As an option we could split the error message and the adb output itself to separate logging calls.